modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/DataSetDisplay.js

```diff
@@ -75,29 +78,30 @@ function DataSetDisplay({
 	sorting: sortingProp,
 	style,
 }) {
+	const {apiURL} = useContext(AppContext);
+
```

Comment:
I'd be *very careful* about adding `useCallback` or `useMemo` just to avoid "re-creating functions repeatedly". Note that in order to make use of either, you are instantiating a new function every time through your render anyway; eg:

```js
```

Reply:
Whoops, accidentally focused the "Comment" button and hit "Enter" before I finished entering the comment. Updated comment incoming...

Reply:
I'd be *very careful* about adding `useCallback` or `useMemo` just to avoid "re-creating functions repeatedly". Note that in order to make use of either, you are instantiating a new function every time through your render anyway; eg: (function chosen at random, but it's just for illustration):

```js
function DataSetDisplay() {
    // stuff

    function updateDataSetItems(dataSetData) {
        setTotal(dataSetData.totalCount);
        updateItems(dataSetData.items);
    }

    // etc
}
```

vs:

```js
function DataSetDisplay() {
    // stuff

    const updateDataSetItems = useCallback((dataSetData) => {
        setTotal(dataSetData.totalCount);
        updateItems(dataSetData.items);
    }, [...]);

    // etc
}
```

The first version is re-instantiating a function object on every render, but so is the second one (the arrow function). In addition, it is paying the overhead of _calling_ `useCallback` on every render, and it is allocating a dependencies array on every render. `useCallback` itself is going to incur the cost of comparing the items in the dependency array (using `Object.is()`), as well as the memory overhead of any related book-keeping details that React is maintaining internally.

Now, I'm not saying "never use `useCallback` or `useMemo`"; I'm just saying, make sure you have a good reason to use them. A non-exhaustive list of good reasons for using them is:

- Use `useMemo` to avoid repeatedly recomputing an intermediate value that is expensive to evaluate (note: in order for this to pay off, the value shouldn't be changing on every render _and_ the cost of determining equality should be cheaper than just redoing the calculation — remember that you're paying the cost of the equality comparison plus the overhead of the function call into React internals plus whatever internal book-keeping React might be doing). In many cases, it's cheaper to just re-do the computation every time.
- Use `useCallback` when you want to maintain a stable identity for a callback function that you will be passing down to children; this can be important when the callback ends up being a member of a `useEffect` dependencies array in the child (or grandchild, or great-grandchild etc). If the callback hasn't really changed, you may find that failure to use `useCallback` causes children to re-run effects unnecessarily because they think something has changed when it really hasn't. This is another context-dependent thing: sometimes you know everything about your children subtree and you can know deterministically that `useCallback` _is_ required; at other times you are writing "library" code and you have no idea what your children are going to be doing, so you may want to err on the side of being defensive and using `useCallback` "just in case".

Another angle on this: https://kentcdodds.com/blog/usememo-and-usecallback/

> _Performance optimizations ALWAYS come with a cost but do NOT always come with a benefit._

Reply:
Made this issue for us to put something about these hooks in the guidelines: https://github.com/liferay/liferay-frontend-projects/issues/383